### PR TITLE
[train] simplify TorchTrainer docstring

### PR DESCRIPTION
**Pull request by matthewdeng**
## Why are these changes needed?

**Generated API Reference:** [TorchTrainer](https://anyscale-ray--38049.com.readthedocs.build/en/38049/train/api/doc/ray.train.torch.TorchTrainer.html)

This PR simplifies the `TorchTrainer` docstring with the following changes:

1. Improve the descriptions of each of the parameters.
2. Reorganize the parameters (moving `dataset_config` immediately after `datasets`).
3. Clean up the example.
4. Reduce the amount of content in the description to make it easier to digest.
    1. Some of this is now captured directly in the parameter descriptions.
    2. Some of it should be linked to user guides.

**Note:** One very unfortunate thing is that the arguments do not get rendered properly, so it's not possible to see the type hints right now. I would like to follow up on this and get this fixed in the future... 
image

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by richardliaw** [comment]
Related to your comment; https://github.com/ray-project/ray/issues/33429

the fix is to not override __new__ in the base trainer

**Comment by matthewdeng** [comment]
Yeah... not clear to me how to though. 😢

**Comment by richardliaw** [comment]
Would it be helpful if I made a draft/closed PR, and perhaps we can find someone to help shepherd a proper fix into 2.7?

**Comment by matthewdeng** [comment]
Yeah that would be great if you have an idea how to - I tried doing something in `BaseTrainer.__init__` but couldn't get it to work since the subclasses don't pass all args in when calling `super(...).__init__(...)`.

File: python/ray/train/torch/torch_trainer.py [hunk 1]
```diff
@@ -235,24 +139,35 @@ def train_loop_per_worker():
...
+            :ref:`Ray Train Loop utilities <train-loop-api>`.
+        train_loop_config: A configuration ``Dict`` to pass in as an argument to
+            ``train_loop_per_worker``.
+            This is typically used for specifying hyperparameters.
+        torch_config: The configuration for setting up the PyTorch Distributed backend.
+            If set to None, a default configuration will be used in which
+            GPU training uses NCCL and CPU training uses Gloo.
+        scaling_config: The configuration for how to scale data parallel training.
+            ``num_workers`` determines how many Python processes are used for training,
+            and ``use_gpu`` determines whether or not each process should use GPUs.
```
Comment (woshiyyya) [inline python/ray/train/torch/torch_trainer.py:155]:
```suggestion
            and ``use_gpu`` determines whether or not each process should use GPUs.
            See :py:class:`ScalingConfig <ray.train.ScalingConfig>` for more info.
```

File: python/ray/train/torch/torch_trainer.py [hunk 2]
```diff
@@ -235,24 +139,35 @@ def train_loop_per_worker():
...
+        run_config: The configuration for the execution of the training run.
```
Comment (woshiyyya) [inline python/ray/train/torch/torch_trainer.py:156]:
```suggestion
        run_config: The configuration for the execution of the training run.
        See :py:class:`RunConfig <ray.train.RunConfig>` for more info.
```

**Review by woshiyyya — APPROVED** [review]
Thank you Matt. This looks great!

File: python/ray/train/torch/torch_trainer.py [hunk 3]
```diff
@@ -16,217 +16,121 @@
...
 
-    This Trainer runs the function ``train_loop_per_worker`` on multiple Ray
-    Actors. These actors already have the necessary torch process group
-    configured for distributed PyTorch training.
+    At a high level, this Trainer does the following:
 
-    The ``train_loop_per_worker`` function is expected to take in either 0 or 1
-    arguments:
+    1. Launches multiple workers as defined by the ``scaling_config``.
+    2. Sets up a PyTorch Distributed environment
```
Comment (angelinalg) [inline python/ray/train/torch/torch_trainer.py:22]: Is `PyTorch Distributed` a product name? Or should it be `PyTorch distributed`?

File: python/ray/train/torch/torch_trainer.py [hunk 4]
```diff
@@ -16,217 +16,121 @@
...
-                # Iterate over epochs and batches
-                for epoch in range(num_epochs):
-                    for batches in dataset_shard.iter_torch_batches(batch_size=32,
-                                dtypes=torch.float):
+                optimizer = torch.optim.SGD(model.parameters(), lr=lr)
 
-                        # Add batch or unsqueeze as an additional dimension [32, x]
-                        inputs, labels = torch.unsqueeze(batches["x"], 1), batches["y"]
-                        output = model(inputs)
+                # Create data loader.
```
Comment (angelinalg) [inline python/ray/train/torch/torch_trainer.py:79]: Does everyone know what a data loader is?

File: python/ray/train/torch/torch_trainer.py (hunk 1 above)
Comment (angelinalg) [reply python/ray/train/torch/torch_trainer.py:155]: Great idea to add the links!

File: python/ray/train/torch/torch_trainer.py [hunk 5]
```diff
@@ -235,24 +139,35 @@ def train_loop_per_worker():
...
+        datasets: The Ray Datasets to ingest for training.
+            Datasets are keyed by name (``{name: dataset}``).
+            Each dataset can be accessed from within the ``train_loop_per_worker``
+            by calling ``ray.train.get_dataset_shard(name)``.
+            Sharding and additional configuration can be done by
+            passing in a ``dataset_config``.
+        dataset_config: The configuration for ingesting the input ``datasets``.
+            By default:
+
+            - The ``"train"`` Dataset will be split equally across workers.
```
Comment (angelinalg) [inline python/ray/train/torch/torch_trainer.py:166]:
```suggestion
            - The ``"train"`` Dataset is split equally across workers.
```

File: python/ray/train/torch/torch_trainer.py [hunk 6]
```diff
@@ -235,24 +139,35 @@ def train_loop_per_worker():
...
+            - All other Datasets will **not** be split.
```
Comment (angelinalg) [inline python/ray/train/torch/torch_trainer.py:167]:
```suggestion
            - All other Datasets are **not** split.
```

File: python/ray/air/config.py [hunk 7]
```diff
@@ -106,7 +106,7 @@ class ScalingConfig:
             argument.
         resources_per_worker: If specified, the resources
             defined in this Dict will be reserved for each worker. The
```
Comment (angelinalg) [inline python/ray/air/config.py:108]:
```suggestion
            defined in this Dict is reserved for each worker.
```

File: python/ray/air/config.py [hunk 8]
```diff
@@ -106,7 +106,7 @@ class ScalingConfig:
...
-            ``CPU`` and ``GPU`` keys (case-sensitive) can be defined to
+            ``"CPU"`` and ``"GPU"`` keys (case-sensitive) can be defined to
```
Comment (angelinalg) [inline python/ray/air/config.py:109]:
```suggestion
            Define the ``"CPU"`` and ``"GPU"`` keys (case-sensitive) to
```

File: python/ray/air/config.py [hunk 9]
```diff
@@ -106,7 +106,7 @@ class ScalingConfig:
...
             override the number of CPU/GPUs used by each worker.
```
Comment (angelinalg) [inline python/ray/air/config.py:110]:
```suggestion
            override the number of CPU or GPUs used by each worker.
```

**Review by angelinalg — APPROVED** [review]
Really, really great improvement. Thank you!

File: python/ray/train/torch/torch_trainer.py [hunk 10]
```diff
@@ -16,217 +16,121 @@
...
-    configured for distributed PyTorch training.
+    At a high level, this Trainer does the following:
 
-    The ``train_loop_per_worker`` function is expected to take in either 0 or 1
-    arguments:
+    1. Launches multiple workers as defined by the ``scaling_config``.
+    2. Sets up a PyTorch Distributed environment
+       on these workers as defined by the ``torch_config``.
+    3. Ingests the input ``datasets`` based on the ``dataset_config``.
+    4. Runs the input ``train_loop_per_worker(train_loop_config)``
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/train/torch/torch_trainer.py:25]: Yeah it is

File: python/ray/train/torch/torch_trainer.py [hunk 11]
```diff
@@ -16,217 +16,121 @@
...
-            import torch.nn as nn
+            from torch import nn
+            from torch.nn.parallel import DistributedDataParallel
 
             import ray
-            from ray import train
             from ray.train import Checkpoint, CheckpointConfig, RunConfig, ScalingConfig
             from ray.train.torch import TorchTrainer
 
             # If using GPUs, set this to True.
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/train/torch/torch_trainer.py:45]: Good catch.

File: python/ray/train/torch/torch_trainer.py (hunk 4 above)
Comment (matthewdeng) [reply python/ray/train/torch/torch_trainer.py:79]: Yeah this is standard PyTorch terminology.

File: python/ray/train/torch/torch_trainer.py [hunk 12]
```diff
@@ -16,217 +16,121 @@
...
+                lr = config["lr"]
+                batch_size = config["batch_size"]
+                num_epochs = config["num_epochs"]
 
-                # Fetch training set
-                dataset_shard = train.get_dataset_shard("train")
+                # Fetch training dataset.
+                train_dataset_shard = ray.train.get_dataset_shard("train")
+
+                # Setup model.
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/train/torch/torch_trainer.py:71]: This is where the model is instantiated, the user can use their own defined model or a pre-defined one. Updated the wording a bit.

File: python/ray/train/torch/torch_trainer.py [hunk 13]
```diff
@@ -16,217 +16,121 @@
...
-            use_gpu = False
-
-            # Define NN layers archicture, epochs, and number of workers
-            input_size = 1
-            layer_size = 32
-            output_size = 1
-            num_epochs = 20
-            num_workers = 3
+            use_gpu = True
+            num_workers = 4
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/train/torch/torch_trainer.py:47]: Added a comment

File: python/ray/train/torch/torch_trainer.py [hunk 14]
```diff
@@ -16,217 +16,121 @@
 class TorchTrainer(DataParallelTrainer):
     """A Trainer for data parallel PyTorch training.
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/train/torch/torch_trainer.py:17]: Yes, but will make this change in a future PR when this becomes the standard way!

**Pull request merged**